operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerEventSource.java

```diff
@@ -11,87 +12,201 @@
+ * resource created/updated using the kubernetes client. (These calls are done automatically by
+ * KubernetesDependentResource implementation.) todo how it works
+ * </p>
+ * <br>
+ * <p>
+ * 2. todo
+ * </p>
+ *
+ * @param <R> resource type watching
+ * @param <P> type of the primary resource
+ */
+public class InformerEventSource<R extends HasMetadata, P extends HasMetadata>
+    extends ManagedInformerEventSource<R, P, InformerConfiguration<R, P>>
+    implements ResourceCache<R>, ResourceEventHandler<R> {
+
+  private static final Logger log = LoggerFactory.getLogger(InformerEventSource.class);
 
-  private final InformerConfiguration<T, P> configuration;
+  private final InformerConfiguration<R, P> configuration;
+  private final EventBuffer<R> eventBuffer = new EventBuffer<>();
 
-  public InformerEventSource(InformerConfiguration<T, P> configuration,
-      EventSourceContext<P> context) {
+  public InformerEventSource(
+      InformerConfiguration<R, P> configuration, EventSourceContext<P> context) {
     super(context.getClient().resources(configuration.getResourceClass()), configuration);
     this.configuration = configuration;
   }
 
-  public InformerEventSource(InformerConfiguration<T, P> configuration,
-      KubernetesClient client) {
+  public InformerEventSource(InformerConfiguration<R, P> configuration, KubernetesClient client) {
     super(client.resources(configuration.getResourceClass()), configuration);
     this.configuration = configuration;
   }
 
   @Override
-  public void onAdd(T t) {
-    propagateEvent(t);
+  public void onAdd(R resource) {
+    onAddOrUpdate("add", resource, () -> InformerEventSource.super.onAdd(resource));
   }
 
   @Override
-  public void onUpdate(T oldObject, T newObject) {
-    if (newObject == null) {
-      // this is a fix for this potential issue with informer:
-      // https://github.com/java-operator-sdk/java-operator-sdk/issues/830
-      propagateEvent(oldObject);
-      return;
-    }
+  public void onUpdate(R oldObject, R newObject) {
+    onAddOrUpdate("update", newObject,
+        () -> InformerEventSource.super.onUpdate(oldObject, newObject));
+  }
 
-    if (oldObject.getMetadata().getResourceVersion()
-        .equals(newObject.getMetadata().getResourceVersion())) {
+  private synchronized void onAddOrUpdate(String operation, R newObject, Runnable superOnOp) {
+    var resourceID = ResourceID.fromResource(newObject);
+    if (eventBuffer.isEventsRecordedFor(resourceID)) {
+      log.info("Recording event for: " + resourceID);
+      eventBuffer.eventReceived(newObject);
       return;
     }
-    propagateEvent(newObject);
+    if (temporalCacheHasResourceWithVersionAs(newObject)) {
+      log.debug(
+          "Skipping event propagation for {}, since was a result of a reconcile action. Resource ID: {}",
+          operation,
+          ResourceID.fromResource(newObject));
+      superOnOp.run();
```

Comment:
`superOnOp` is run in both branches so could be run outside of the if/else. No big deal either way. :)